### PR TITLE
non-bfb changes for shoc_assumed_pdf porting

### PR DESCRIPTION
**Pull request by tcclevenger**
Several changes to `bfb_*()` function for `shoc_assumed_pdf` porting.

All changes to `bfb_square()` and `bfb_sqrt()` cause no change to output from bfb testing, however, like other bfb changes, `bfb_cube` and `bfb_pow` can change the `shoc_run_and_cmp` output by a significant amount (~1e-1, `host_dse` is always the largest).

Again, I test all these by calculating the relative difference between variables containing `bfb_cube(x)` and those containing `x**3` everywhere it is used during ./shoc_run_and_cmp (likewise with `bfb_pow`). For all of these changes, the maximum relative difference of any one change is
`abs(old_val - new_val)/abs(old_val) < 7e-06(SP), 3e-14 (DP)`

### CODE REVIEW

File: components/cam/src/physics/cam/shoc.F90 [hunk 1]
```diff
@@ -1993,14 +1993,15 @@ subroutine shoc_assumed_pdf(&
...
+  real(rtype) :: sqrt2, sqrtpi
+  sqrt2=bfb_sqrt(2._rtype)
+  sqrtpi=bfb_sqrt(2._rtype*3.14_rtype)
 
   epsterm=rgas/rv
 
   thl_tol=1.e-2_rtype
   rt_tol=1.e-4_rtype
-  w_tol_sqd=(2.e-2_rtype)**2
+  w_tol_sqd=bfb_square(2.e-2_rtype)
```
Comment (tcclevenger) [inline components/cam/src/physics/cam/shoc.F90:2004]: Should `2.e` be changed to `2._rtype`?

File: components/cam/src/physics/cam/shoc.F90 [hunk 2]
```diff
@@ -2190,9 +2191,10 @@ subroutine shoc_assumed_pdf_vv_parameters(&
   real(rtype) :: sqrtw2t
 
   ! parameters
-  real(rtype), parameter :: w_tol_sqd=(2.e-2_rtype)**2
+  real(rtype) :: w_tol_sqd
+  w_tol_sqd = bfb_square(2.e-2_rtype)
 
-  Skew_w=w3var/w_sec**(3./2.)
+  Skew_w=w3var/bfb_pow(w_sec,(3./2.))
```
Comment (tcclevenger) [inline components/cam/src/physics/cam/shoc.F90:2197]: Should `3./2.` be changed to `3._rtype/2._rtype`?

**Review by tcclevenger — COMMENTED** [review]
I've left 2 questions about other potential changes..

File: components/cam/src/physics/cam/shoc.F90 (hunk 1 above)
Comment (ambrad) [reply components/cam/src/physics/cam/shoc.F90:2004]: Yes.

File: components/cam/src/physics/cam/shoc.F90 [hunk 3]
```diff
@@ -1993,14 +1993,15 @@ subroutine shoc_assumed_pdf(&
   real(rtype) :: qw_sec_zt(shcol,nlev)
 
   ! define these so they don't have to be computed more than once
-  real(rtype), parameter :: sqrt2 = sqrt(2._rtype)
-  real(rtype), parameter :: sqrtpi = sqrt(2._rtype*3.14_rtype)
+  real(rtype) :: sqrt2, sqrtpi
+  sqrt2=bfb_sqrt(2._rtype)
+  sqrtpi=bfb_sqrt(2._rtype*3.14_rtype)
```
Comment (ambrad) [inline components/cam/src/physics/cam/shoc.F90:1998]: @bogensch would it make sense in a separate PR to use pi to full precision here?

File: components/cam/src/physics/cam/shoc.F90 (hunk 2 above)
Comment (ambrad) [reply components/cam/src/physics/cam/shoc.F90:2197]: Better: `bfb_sqrt(bfb_cube(w_sec))`.

**Review by ambrad — CHANGES_REQUESTED** [review]

File: components/cam/src/physics/cam/shoc.F90 (hunk 3 above)
Comment (bartgol) [reply components/cam/src/physics/cam/shoc.F90:1998]: Also, is this meant to be sqrt(pi) or is it really sqrt(2*pi)?

Comment (bogensch) [reply components/cam/src/physics/cam/shoc.F90:1998]: It is meant to be sqrt(2*pi)... and yes, I agree that full precision pi should be used.

File: components/cam/src/physics/cam/shoc.F90 (hunk 1 above)
Comment (tcclevenger) [reply components/cam/src/physics/cam/shoc.F90:2004]: Actually, I think I read this incorrect. This is `2e-2`.

File: components/cam/src/physics/cam/shoc.F90 (hunk 3 above)
Comment (tcclevenger) [reply components/cam/src/physics/cam/shoc.F90:1998]: I'll create one.

Comment (PeterCaldwell) [reply components/cam/src/physics/cam/shoc.F90:1998]: Could we please use the name `sqrt2pi` instead of `sqrtpi` for `sqrt(2*pi)` ? Otherwise the name is misleading.

Comment (tcclevenger) [reply components/cam/src/physics/cam/shoc.F90:1998]: In  `physics_constants.hpp`, pi is defined `pi=3.14159265`. Shouldn't it be more digits (in both places)? Should I add a constant to `shoc.F90` similarly?

Comment (bartgol) [reply components/cam/src/physics/cam/shoc.F90:1998]: Since pi is a universal constant, I would put it in `physics/share` at the very least. It might be actually belong into a more common package (see E3SM-Project/EKAT#48 as well).

Either way: why forcing full precision? This is F90, so it probably doesn't matter, but if the same design stays in C++, isn't there the chance that it will cause all downstream calculation to be in double precision? It seems to me that, unless there's a clear _need_ for pi in double precision, then its precision should match that of the rest of the library...

Comment (tcclevenger) [reply components/cam/src/physics/cam/shoc.F90:1998]: Here is a [PR for adding pi to shoc.F90](https://github.com/E3SM-Project/scream/pull/677) (instead of using 3.14) and changing the name to `sqrt2pi`.

**Comment by tcclevenger** [comment]
@bartgol @ambrad This branch is now rebased to master and the failing tests are expected. See initial comment on the PR for the rel_diff of the changes/

**Review by ambrad — APPROVED** [review]

**Comment by ambrad** [comment]
@tcclevenger should I manually merge?

**Comment by tcclevenger** [comment]
@ambrad yes
